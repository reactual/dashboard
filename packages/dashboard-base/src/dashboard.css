body {
    margin: 0;
    padding: 0;
    font-family: sans-serif;
}

dl dt {
    margin-top: 1em;
    font-weight: bold;
}

dl dd {
    margin: 0.1em 0 0 0;
}

.header {
    padding: 0.5em 0 0.5em 0;
    background-color: black;
    color: white;
}

.header ul {
    float: right;
}

.header ul li {
    display: inline-block;
    list-style: none;
    padding: 0 0 0 1em;
    vertical-align: middle;
}

.header .ms-Button,
.header a {
    color: white !important;
}

.header a:hover,
.header .ms-Button:hover {
    color: #0078d7 !important;
}

.header a {
    font-weight: bold;
    text-decoration: none;
}
