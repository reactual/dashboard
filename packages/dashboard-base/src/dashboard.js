import Immutable from "immutable"
import React, { Component } from "react"
import { Provider, connect } from "react-redux"
import { Router, IndexRoute, Route, Redirect, Link, browserHistory } from "react-router"

import "./dashboard.css"
import logo from "./logo.svg"
import GetStarted from "./get-started"
import { createReduxStore } from "./store"
import { updateSelectedResource } from "./router"
import { ActivityMonitor, monitorActivity } from "./activity-monitor"
import { NotificationBar, watchForError } from "./notifications"
import { LoginForm, UserAccount, faunaClient } from "./authentication"
import { IntercomWidget } from "./external/intercom"
import { KeyType } from "./persistence/faunadb-wrapper"
import { ToggleRepl } from "./repl"
import { Events } from "./plugins"

import {
  NavTree,
  DatabaseForm,
  ClassForm,
  ClassManager,
  IndexForm,
  IndexManager,
  KeysManager,
  loadSchemaTree
} from "./schema"

export class Container extends Component {

  componentWillReceiveProps(next) {
    const oldParams = Immutable.Map(this.props.params)
    const newParams = Immutable.Map(next.params)

    if (this.props.faunaClient !== next.faunaClient || !oldParams.equals(newParams)) {
      this.updateSelectedResource(
        next.faunaClient,
        next.params
      )
    }
  }

  updateSelectedResource(client, params) {
    if (!client) return

    this.props.dispatch(
      monitorActivity(
        watchForError(
          "Unexpected error while loading database. It may not exist or your key can't access it",
          (dispatch) => {
            const selected = dispatch(updateSelectedResource(params))
            return dispatch(loadSchemaTree(client, selected.get("database")))
          }
        )
      )
    )
  }

  render() {
    return <div className="ms-Fabric ms-font-m">
        <ToggleRepl>
          <div className="ms-Grid">
            <div className="ms-Grid-row header">
              <div className="ms-Grid-col ms-u-sm5 ms-u-md6 ms-u-lg3 ms-u-xl2">
                <Link to="/"><img src={logo} alt="logo" /></Link>
              </div>
              <div className="ms-Grid-col ms-u-sm12 ms-u-md6 ms-u-lg9 ms-u-xl10">
                <ul>
                  <li><ActivityMonitor /></li>
                  <li><a href="http://fauna.com/tutorials" target="_blank" rel="noopener noreferrer">Tutorials</a></li>
                  <li><a href="http://fauna.com/documentation" target="_blank" rel="noopener noreferrer">Documentation</a></li>
                  <li><a href="https://fauna.com/resources#drivers" target="_blank" rel="noopener noreferrer">Drivers</a></li>
                  <li><UserAccount /></li>
                </ul>
              </div>
            </div>
            { this.props.faunaClient ? this.renderMainView() : null }
          </div>
        </ToggleRepl>
        <NotificationBar />
        <LoginForm />
        <IntercomWidget />
      </div>
  }

  renderMainView() {
    if (this.props.faunaClient.hasPrivileges(KeyType.ADMIN)) {
      return <div className="ms-Grid-row">
        <div className="ms-Grid-col ms-u-sm12 ms-u-md6 ms-u-lg4">
          <NavTree />
        </div>
        <div className="ms-Grid-col ms-u-sm12 ms-u-md6 ms-u-lg8">
          {this.props.children}
        </div>
      </div>
    }

    return <div className="ms-Grid-row">
      <div className="ms-Grid-col ms-u-sm12 ms-u-md3 ms-u-lg2">
        <NavTree />
      </div>
      <div className="ms-Grid-col ms-u-sm12 ms-u-md9 ms-u-lg10">
        {this.props.children}
      </div>
    </div>
  }
}

export default class Dashboard extends Component {

  static Container = connect(
    state => ({
      faunaClient: faunaClient(state)
    })
  )(Container)

  static NotFound = () => <h1>404.. This page is not found!</h1>
  static isProduction = process.env.NODE_ENV === "production"

  constructor(props) {
    super(props)
    this.store = createReduxStore()
    this.firePageChanged = this.firePageChanged.bind(this)
  }

  firePageChanged() {
    Events.fire("@@dashboard/page-changed", {
      pathname: window.location.pathname
    })
  }

  render() {
    return <Provider store={this.store}>
        <Router onUpdate={this.firePageChanged} history={browserHistory}>
          <Redirect from="/" to="/db" />
          <Route path="/db" component={Dashboard.Container}>
            <Route path="indexes/:indexName" component={IndexManager} />
            <Route path="indexes" component={IndexForm} />
            <Route path="classes/:className" component={ClassManager} />
            <Route path="classes" component={ClassForm} />
            <Route path="databases" component={DatabaseForm} />
            <Route path="keys" component={KeysManager} />
            <Route path="**/indexes/:indexName" component={IndexManager} />
            <Route path="**/indexes" component={IndexForm} />
            <Route path="**/classes/:className" component={ClassManager} />
            <Route path="**/classes" component={ClassForm} />
            <Route path="**/databases" component={DatabaseForm} />
            <Route path="**/keys" component={KeysManager} />
            <IndexRoute component={GetStarted} />
          </Route>
          <Route path="*" component={Dashboard.NotFound} />
        </Router>
      </Provider>
  }
}
