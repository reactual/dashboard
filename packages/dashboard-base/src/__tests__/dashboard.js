import React from "react"
import { shallow } from "enzyme"
import { shallowToJson } from "enzyme-to-json"

import { Container } from "../dashboard"

describe("Container Component", () => {
  let comp, client

  beforeEach(() => {
    client = { hasPrivileges: jest.fn(() => true) }
    comp = shallow(<Container client={client} />)
  })

  it("displays the base layout", () => {
    expect(shallowToJson(comp)).toMatchSnapshot()
  })

  it("renders a smaller sidebar when using a server key", () => {
    client.hasPrivileges.mockReturnValue(false)
    expect(shallowToJson(comp)).toMatchSnapshot()
  })
})
