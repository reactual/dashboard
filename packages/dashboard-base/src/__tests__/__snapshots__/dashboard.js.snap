// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Container Component displays the base layout 1`] = `
<div
  className="ms-Fabric ms-font-m"
>
  <Connect(ToggleRepl)>
    <div
      className="ms-Grid"
    >
      <div
        className="ms-Grid-row header"
      >
        <div
          className="ms-Grid-col ms-u-sm5 ms-u-md6 ms-u-lg3 ms-u-xl2"
        >
          <Link
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/"
          >
            <img
              alt="logo"
              src="logo.svg"
            />
          </Link>
        </div>
        <div
          className="ms-Grid-col ms-u-sm12 ms-u-md6 ms-u-lg9 ms-u-xl10"
        >
          <ul>
            <li>
              <Connect(ActivityMonitor) />
            </li>
            <li>
              <a
                href="http://fauna.com/tutorials"
                rel="noopener noreferrer"
                target="_blank"
              >
                Tutorials
              </a>
            </li>
            <li>
              <a
                href="http://fauna.com/documentation"
                rel="noopener noreferrer"
                target="_blank"
              >
                Documentation
              </a>
            </li>
            <li>
              <a
                href="https://fauna.com/resources#drivers"
                rel="noopener noreferrer"
                target="_blank"
              >
                Drivers
              </a>
            </li>
            <li>
              <Connect(UserAccount) />
            </li>
          </ul>
        </div>
      </div>
    </div>
  </Connect(ToggleRepl)>
  <Connect(NotificationBar) />
  <Connect(LoginForm) />
  <Connect(IntercomWidget) />
</div>
`;

exports[`Container Component renders a smaller sidebar when using a server key 1`] = `
<div
  className="ms-Fabric ms-font-m"
>
  <Connect(ToggleRepl)>
    <div
      className="ms-Grid"
    >
      <div
        className="ms-Grid-row header"
      >
        <div
          className="ms-Grid-col ms-u-sm5 ms-u-md6 ms-u-lg3 ms-u-xl2"
        >
          <Link
            onlyActiveOnIndex={false}
            style={Object {}}
            to="/"
          >
            <img
              alt="logo"
              src="logo.svg"
            />
          </Link>
        </div>
        <div
          className="ms-Grid-col ms-u-sm12 ms-u-md6 ms-u-lg9 ms-u-xl10"
        >
          <ul>
            <li>
              <Connect(ActivityMonitor) />
            </li>
            <li>
              <a
                href="http://fauna.com/tutorials"
                rel="noopener noreferrer"
                target="_blank"
              >
                Tutorials
              </a>
            </li>
            <li>
              <a
                href="http://fauna.com/documentation"
                rel="noopener noreferrer"
                target="_blank"
              >
                Documentation
              </a>
            </li>
            <li>
              <a
                href="https://fauna.com/resources#drivers"
                rel="noopener noreferrer"
                target="_blank"
              >
                Drivers
              </a>
            </li>
            <li>
              <Connect(UserAccount) />
            </li>
          </ul>
        </div>
      </div>
    </div>
  </Connect(ToggleRepl)>
  <Connect(NotificationBar) />
  <Connect(LoginForm) />
  <Connect(IntercomWidget) />
</div>
`;
